controllers/drain_controller.go

```diff
@@ -0,0 +1,465 @@
+	drainingNodes := len(mcpPauseDrainNodeList.Items) + len(DrainingNodeList.Items)
+	reqLogger.Info("Count of draining", "drainingNodes", drainingNodes)
+
+	// if both are Idle we don't need to do anything
+	if nodeDrainAnnotation == constants.DrainIdle &&
+		nodeStateDrainAnnotationCurrent == constants.DrainIdle {
+		reqLogger.Info("node and nodeState are on idle nothing todo")
+		return reconcile.Result{}, nil
+	}
+
+	if nodeDrainAnnotation == constants.DrainIdle &&
+		nodeStateDrainAnnotationCurrent == constants.DrainComplete {
+		err = dr.completeDrain(ctx, node)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		err = utils.AnnotateObject(nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.DrainIdle, dr.Client)
+		return ctrl.Result{}, err
+	}
+
+	// the node request to drain, but we are at the limit so we re-enqueue the request
+	if drainingNodes >= maxParallelNodeConfiguration {
+		reqLogger.Info("MaxParallelNodeConfiguration limit reached for draining nodes re-enqueue the request")
+		// TODO: make this time configurable
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	if nodeDrainAnnotation == constants.DrainRequired &&
+		nodeStateDrainAnnotationCurrent == constants.DrainIdle {
+		// pause MCP if we are on openshift but not hypershift deployment method as there is MCP in hypershift
+		if !dr.openshiftContext.IsOpenshiftCluster() || (dr.openshiftContext.IsOpenshiftCluster() && dr.openshiftContext.IsHypershift()) {
+			err = utils.AnnotateObject(nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.Draining, dr.Client)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			nodeStateDrainAnnotationCurrent = constants.Draining
+		} else if dr.openshiftContext.IsOpenshiftCluster() && !dr.openshiftContext.IsHypershift() {
+			nodePoolName, err := dr.openshiftContext.GetNodeMachinePoolName(node)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+
+			if err := dr.pauseMCP(nodeNetworkState, nodePoolName); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			nodeStateDrainAnnotationCurrent = nodeNetworkState.Annotations[constants.NodeStateDrainAnnotationCurrent]
+		}
+	}
+
+	// the node is MCP paused already, so we need to continue from here
+	// TODO: with parallel draining we need to also check that there is no other reconcile loop that is working
+	if nodeDrainAnnotation == constants.DrainRequired &&
+		(nodeStateDrainAnnotationCurrent == constants.DrainMcpPaused || nodeStateDrainAnnotationCurrent == constants.Draining) {
+		err = dr.drainNode(ctx, node)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		err = utils.AnnotateObject(nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.DrainComplete, dr.Client)
+		return ctrl.Result{}, err
+	}
+
+	if nodeDrainAnnotation == constants.DrainIdle &&
+		nodeStateDrainAnnotationCurrent == constants.DrainMcpPaused {
```

Comment:
what if nodeStateDrainAnnotationCurrent == constants.Draining ?
i think we should also completeDrain